src/python/pants/backend/python/rules/targets.py

```diff
@@ -0,0 +1,207 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from pathlib import PurePath
+from typing import Any, ClassVar, Optional
+
+from pants.build_graph.address import Address
+from pants.engine.objects import union
+from pants.engine.target import (
+    COMMON_TARGET_FIELDS,
+    BoolField,
+    PrimitiveField,
+    Sources,
+    SourcesResult,
+    StringField,
+    StringOrStringListField,
+    Target,
+)
+
+
+@union
+class PythonSources(Sources):
+    @classmethod
+    def validate_result(cls, result: SourcesResult) -> None:
+        non_python_files = [fp for fp in result.snapshot.files if not PurePath(fp).suffix == ".py"]
+        if non_python_files:
+            raise ValueError(
+                f"Target {result.address} has non-Python sources in its `sources` field: "
+                f"{non_python_files}"
+            )
+
+
+class PythonLibrarySources(PythonSources):
+    default_globs: ClassVar = ("*.py", "!test_*.py", "!*_test.py", "!conftest.py")
+
+
+class PythonTestsSources(PythonSources):
+    default_globs: ClassVar = ("test_*.py", "*_test.py", "conftest.py")
+
+
+class PythonBinarySources(PythonSources):
+    @classmethod
+    def validate_result(cls, result: SourcesResult) -> None:
+        super().validate_result(result)
+        if len(result.snapshot.files) not in [0, 1]:
+            raise ValueError(
+                "Binary targets must have only 0 or 1 source files. Any additional files should "
+                "be put in a `python_library` which is added to `dependencies`"
+            )
+
+
+class Compatibility(StringOrStringListField):
+    """A string for Python interpreter constraints on this target.
+
+    This should be written in Requirement-style format, e.g. `CPython==2.7.*` or `CPython>=3.6,<4`.
+
+    As a shortcut, you can leave off `CPython`, e.g. `>=2.7` will be expanded to `CPython>=2.7`.
+    """
+
+    alias: ClassVar = "compatibility"
+
+
+# TODO: Deal with the `provides` field. This will at least allow us to correctly parse the valid,
```

Comment:
"parse the valid"?

Reply:
Oh oops. Will fix.

Reply:
Actually, going to do in a follow up to save CI resources.